commit 38fa2c750335e89c384823d592c0ca6457eecea2
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Mon Dec 4 16:37:43 2023 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Mon Dec 4 16:37:43 2023 -0500

    Fix meta file processing in storage and improve schedule job retrieval (#2186)
    
    * fix meta file processing in storage; enhance shedule job
    
    * fix test cases
    
    * remove unused import
    
    * use mark file to reduce meta reading
    
    * improve type hints
    
    * rename func and improve docstring
    
    * make get_jobs_to_schedule abstract
---
 nvflare/private/fed/server/job_runner.py | 19 ++++++++-----------
 1 file changed, 8 insertions(+), 11 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 48154cad..ae2cb7d5 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -395,10 +395,15 @@ class JobRunner(FLComponent):
             thread.start()
 
             while not self.ask_to_stop:
+                time.sleep(1.0)
                 if not isinstance(engine.server.server_state, HotState):
-                    time.sleep(1.0)
                     continue
-                approved_jobs = job_manager.get_jobs_by_status(RunStatus.SUBMITTED, fl_ctx)
+
+                if not engine.get_clients():
+                    # no clients registered yet - don't try to schedule!
+                    continue
+
+                approved_jobs = job_manager.get_jobs_to_schedule(fl_ctx)
                 self.log_debug(
                     fl_ctx, f"{fl_ctx.get_identity_name()} Got approved_jobs: {approved_jobs} from the job_manager"
                 )
@@ -477,9 +482,6 @@ class JobRunner(FLComponent):
                             self.log_error(
                                 fl_ctx, f"Failed to run the Job ({ready_job.job_id}): {secure_format_exception(e)}"
                             )
-
-                time.sleep(1.0)
-
             thread.join()
         else:
             self.log_error(fl_ctx, "There's no Job Manager defined. Won't be able to run the jobs.")
@@ -543,12 +545,7 @@ class JobRunner(FLComponent):
 
     @staticmethod
     def _get_all_running_jobs(job_manager, fl_ctx):
-        all_jobs = []
-        dispatched_jobs = job_manager.get_jobs_by_status(RunStatus.DISPATCHED, fl_ctx)
-        all_jobs.extend(dispatched_jobs)
-        running_jobs = job_manager.get_jobs_by_status(RunStatus.RUNNING, fl_ctx)
-        all_jobs.extend(running_jobs)
-        return all_jobs
+        return job_manager.get_jobs_by_status([RunStatus.RUNNING, RunStatus.DISPATCHED], fl_ctx)
 
     def stop_run(self, job_id: str, fl_ctx: FLContext):
         engine = fl_ctx.get_engine()
